compiler/optimizer/InductionVariable.cpp

```diff
@@ -4946,70 +4978,123 @@ TR::Node *TR_LoopStrider::signExtend(
                child->getOpCodeValue() == TR::lload,
                "exload l2i n%un's child should be an lload\n",
                node->getGlobalIndex());
-            ext = child;
+            result.extended = child;
+            result.cancelsExt = true;
+            result.cancelsTrunc = true;
             }
          break;
 
       case TR::iadd:
-         ext = signExtendBinOp(TR::ladd, node, exLoads, memo);
+         result = signExtendBinOp(TR::ladd, node, exLoads, memo);
          break;
 
       case TR::isub:
-         ext = signExtendBinOp(TR::lsub, node, exLoads, memo);
+         result = signExtendBinOp(TR::lsub, node, exLoads, memo);
          break;
 
       case TR::imul:
-         ext = signExtendBinOp(TR::lmul, node, exLoads, memo);
+         result = signExtendBinOp(TR::lmul, node, exLoads, memo);
          break;
 
       default:
-         break; // just don't set ext
+         break; // just don't populate result
       }
 
-   if (ext != NULL)
+   if (result.extended != NULL)
       {
-      memo.insert(std::make_pair(node->getGlobalIndex(), ext));
+      result.extended->incReferenceCount();
+      memo.insert(std::make_pair(node->getGlobalIndex(), result));
       if (trace())
          traceMsg(comp(),
             "[Sign-Extn] sign-extended n%un %s into n%un %s\n",
             node->getGlobalIndex(),
             node->getOpCode().getName(),
-            ext->getGlobalIndex(),
-            ext->getOpCode().getName());
+            result.extended->getGlobalIndex(),
+            result.extended->getOpCode().getName());
       }
 
-   return ext;
+   TR_ASSERT(
+      (result.extended != NULL) == (result.cancelsExt || result.cancelsTrunc),
+      "n%un %s should extend if and only if a conversion is canceled\n",
+      node->getGlobalIndex(),
+      node->getOpCode().getName());
+
+   return result;
    }
 
 // part of implementation of signExtend (above)
-TR::Node *TR_LoopStrider::signExtendBinOp(
+TR_LoopStrider::SignExtEntry TR_LoopStrider::signExtendBinOp(
    TR::ILOpCodes op64,
    TR::Node *node,
    TR::NodeChecklist &exLoads,
-   NodeMap &memo)
+   SignExtMemo &memo)
    {
+   const SignExtEntry rejected;
    if (!node->cannotOverflow())
-      return NULL;
-
-   TR::Node *extLeft = signExtend(node->getChild(0), exLoads, memo);
+      return rejected;
+
+   static const char * const moveDisableEnv =
+      feGetEnv("TR_disableLoopStriderMoveSignExtIntoChild");
+   static const bool moveDisable =
+      moveDisableEnv != NULL && moveDisableEnv[0] != '\0';
```

Comment:
I was under the impression that it's rare (in general) to treat unset environment variables differently from those set to the empty string. For example, shell scripts often test for environment variables with something like `test -n "$VAR"`. Maybe this is a peculiarity of my own, but if the code tested only `moveDisableEnv != NULL`, I would be surprised at the behaviour of a script like this:

    for disable in '' 1; do
        TR_disableLoopStriderMoveSignExtIntoChild="$disable" vm-with-jit [...]
    done

It also isn't entirely without precedent. I've introduced multiple other environment variable tests following the same convention, even just in this file. Most recently, we have `TR_disableLoopStriderWidenComparison` from #662.

Of course, it's not as though this convention helps all that much to prevent error. Just about anybody might still be surprised if the above loop said, for example, `for disable in 0 1`, or `for disable in no yes`. And nothing prevents a typo in the variable name.

For that reason, this should ideally be a JIT option instead, if we want to be particular about it. The name would be checked, and it could be restricted to an option set, and so on. I used an environment variable here for expediency, because I often get merge conflicts when I'm adding a JIT option. (I think this is the story of most of our environment variables.)

So with that said, at your preference I can change this environment variable to use the prevailing convention of `moveDisable = moveDisableEnv != NULL` (full-stop), or make it into a JIT option, or leave it as is. You should just be aware that there are others like this one. I hope nobody is out there with the opposite of my expectation, trying to disable things with `TR_disableSomething=''`.